lib/graphql/query/context.rb

```diff
@@ -163,15 +164,30 @@ def initialize(query:, values: , object:)
       # @api private
       attr_writer :value
 
-      def_delegators :@provided_values, :[], :[]=, :to_h, :to_hash, :key?, :fetch, :dig
-      def_delegators :@query, :trace, :interpreter?
+      # @api private
+      attr_accessor :scoped_context
 
-      # @!method [](key)
-      #   Lookup `key` from the hash passed to {Schema#execute} as `context:`
+      def_delegators :@provided_values, :[]=
+      def_delegators :to_h, :fetch, :dig
+      def_delegators :@query, :trace, :interpreter?
 
       # @!method []=(key, value)
       #   Reassign `key` to the hash passed to {Schema#execute} as `context:`
 
+      # Lookup `key` from the hash passed to {Schema#execute} as `context:`
+      def [](key)
+        return @scoped_context[key] if @scoped_context.key?(key)
+        @provided_values[key]
+      end
+
+      def to_h
+        @provided_values.merge(@scoped_context)
```

Comment:
Should this mutate with `merge!`?

Reply:
Nope! This needs to create a new object, because lazy resolvers being executed out-of-order still need an untainted reference to the `scoped_context` they had originally. If this proves to be allocating too many objects or consuming too much memory, we should consider using an immutable data structure with structural sharing instead of `Hash` here.